psilbhb/stim/wav_set.py

```diff
@@ -980,3 +1035,483 @@ def score_response(self, outcome, repeat_incorrect=True, trial_idx=None):
+            TODO: add fraction catch_frequency of trials with no target on top of
+            regular trial set
+        :param fg_delay: float
+        :param response_window: {tuple, np.array, None}
+            None converts to (0,1) = 0 to 1 sec after fg starts playing
+            If array, length should bg >= FgSet.max_index
+        :param random_seed: int
+        """
+
+        if CatchFGSet is None:
+            if FgSet is None:
+                self.FgSet = WaveformSet()
+            else:
+                self.FgSet = FgSet
+        else:
+            if FgSet is None:
+                raise ValueError(f"Must specify FgSet if CatchFGSet is specified")
+            else:
+                self.FgSet = cat_MCWavFileSets(FgSet, CatchFGSet, frac_set1=CatchFG_frac)
+
+        if CatchBGSet is None:
+            if BgSet is None:
+                self.BgSet = WaveformSet()
+            else:
+                self.BgSet = BgSet
+        else:
+            if BgSet is None:
+                raise ValueError(f"Must specify BgSet if CatchBGSet is specified")
+            else:
+                self.BgSet = cat_MCWavFileSets(BgSet, CatchBGSet, frac_set1=CatchBG_frac)
+
+        self.combinations = combinations
+        self.fg_switch_channels = fg_switch_channels
+        self.bg_switch_channels = bg_switch_channels
+        self.catch_frequency = catch_frequency
+        self._fg_delay = fg_delay
+        self._fg_snr = fg_snr
+        self.primary_channel = primary_channel
+
+        #region migrate code
+        self.migrate_fraction = migrate_fraction
+        self.migrate_start = migrate_start
+        self.migrate_stop = migrate_stop
+        #endregion
+
+        if response_window is None:
+            self.response_window = (0.0, 1.0)
+        else:
+            self.response_window = response_window
+
+        self.random_seed = random_seed
+        self.current_trial_idx = -1
+
+        # trial management
+        self.trial_wav_idx = np.array([], dtype=int)
+        self.trial_outcomes = np.array([], dtype=int)
+        self.trial_is_repeat = np.array([], dtype=int)
+        self.current_full_rep = 0
+
+        self.update()
+
+    @property
+    def wav_per_rep(self):
+        return np.min([len(self.bg_index), len(self.fg_index)])
+
+    def update(self, trial_idx=None):
+        """figure out indexing to map wav_set idx to specific members of FgSet and BgSet.
+        manage trials separately to allow for repeats, etc."""
+        _rng = np.random.RandomState(self.random_seed)
+
+        bg_range = np.arange(self.BgSet.max_index, dtype=int)
+        fg_range = np.arange(self.FgSet.max_index, dtype=int)
+
+        if self.fg_switch_channels:
+            fg_channel = np.concatenate((np.zeros_like(fg_range), np.ones_like(fg_range)))
+            fg_range = np.tile(fg_range, 2)
+        else:
+            fg_channel = np.zeros_like(fg_range) + self.primary_channel
+
+        if self.bg_switch_channels == False:
```

Comment:
Don't switch between datatypes. Keep them the same (i.e., call it `"no"` or `"false"` instead of `False`). The other possible values for this attribute are strings so be consistent with that.